File: Problem 004/solution.go
package main

import "fmt"

func isPalindrom(num int) bool {
    reversed := 0
    number := num

    for num > 0 {
        reversed = reversed * 10 + num % 10
        num /= 10
    }

    return reversed == number
}

func main() {
    var t int
    fmt.Scan(&t)
    for i := 0; i < t; i++ {
        var n, prod int
        fmt.Scan(&n)
        largest_prime := -1

        for j := 990; j <= 100; j-=11 {
            for k := 999; k <= 100; k-- {
                prod = j * k
                if prod > largest_prime && isPalindrom(prod) && prod < n {
                    largest_prime = prod
                    break
                } else if prod < largest_prime {
                    break
                }
            }
        }
        fmt.Println(largest_prime)
    }
}

File: Problem 004/solution.c
#include <stdio.h>

int isPalindrom(int num) {
    int reversed = 0, number = num;

    while (num > 0) {
        reversed = reversed * 10 + num % 10;
        num /= 10;
    }

    return (reversed == number);
}

int main(int argc, char const *argv[]) {
    int test;
    scanf("%d", &test);
    for (int c = 0; c < test; c++) {
        int n;
        scanf("%d", &n);
        long long int prod, largestPrime = -1;

        for (int i = 990; i > 99; i-=11) {
            for (int j = 999; j > 99; j--) {
                prod = i * j;
                if ((prod > largestPrime) && (isPalindrom(prod)) && (prod < n)) {
                    largestPrime = prod;
                    break;
                }
                else if (prod < largestPrime)
                    break;
            }
        }
        printf("%lld\n", largestPrime);
    }
    return 0;
}
